### PR TITLE
Add support for ternary const entries

### PR DESCRIPTION
**Pull request by kmateuszssak**
This PR adds possibility to use const entries with ternary tables.

### CODE REVIEW

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 1]
```diff
@@ -564,4 +579,344 @@ bool EBPFTablePSA::dropOnNoMatchingEntryFound() const {
...
+                              keysArray,
+                              valuesArray);
+        builder->endOfStatement(true);
+
+        tuple_id++;
+    }
+}
+
+void EBPFTablePSA::emitKeysAndValues(CodeBuilder *builder,
+                                            std::vector<const IR::Entry *> &samePrefixEntries,
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:673]: inconsistent indentation

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 2]
```diff
@@ -564,4 +579,344 @@ bool EBPFTablePSA::dropOnNoMatchingEntryFound() const {
...
+                expr->apply(cg);
+                builder->append(" & ");
+                unsigned width = 0;
+                if (auto hasWidth = ebpfType->to<IHasWidth>()) {
+                    width = hasWidth->widthInBits();
+                } else {
+                    BUG("Cannot assess field bit width");
+                }
+                builder->append("0x");
+                for (int j = 0; j < width / 8; j++)
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:712]: does this work with non-aligned widths too?

Comment (mbudiu-vmw) [reply backends/ebpf/psa/ebpfPsaTable.cpp:712]: you should factor this into a utility function getHexMask, or something of the kind.

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 3]
```diff
@@ -564,4 +579,344 @@ bool EBPFTablePSA::dropOnNoMatchingEntryFound() const {
...
+    } else {
+        builder->appendFormat("%s.next_tuple_mask = %s", valueMask, nextMask);
+        builder->endOfStatement(true);
+        builder->emitIndent();
+        builder->appendFormat("%s.has_next = 1", valueMask);
+        builder->endOfStatement(true);
+    }
+}
+
+std::vector<std::vector<const IR::Entry*>> EBPFTablePSA::getConstEntriesGroupedByPrefix() {
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:803]: A small comment would help make this function easier to interpret.

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 4]
```diff
@@ -564,4 +579,344 @@ bool EBPFTablePSA::dropOnNoMatchingEntryFound() const {
...
+    std::vector<std::vector<const IR::Entry*>> entriesGroupedByPrefix;
+    const IR::EntriesList* entries = table->container->getEntries();
+    if (entries != nullptr) {
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:806]: especially for long ifs I find it simpler to `if (!entries) return ...;`
One less level of indentation afterwards.

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by kmateuszssak** [comment]
@mbudiu-vmw I've addressed comments.

**Pull request merged**